# **Exploratory Data Analisys: Companies**

In the notebook dedicated to the exploratory data analysis (EDA) of companies listed on B3, the filters are specifically designed to meet the needs of investors with a Buy and Hold strategy. These filters allow for better visualization and interpretation of data, facilitating the identification of companies that offer good long-term appreciation and stability, which is essential for this type of investment.

Here are the fundamental metrics that are used to filter and assess the companies:

- **`market_cap`**: to filter companies by their market capitalization, allowing the identification of companies with greater market stability and solidity.
- **`dividend_rate`**: to select companies based on the dividend rate paid, which can be an indicator of consistent cash generation.
- **`trailing_annual_dividend_yield`**: for companies based on past annual dividend yields, a metric that reflects profitability over the stock price.
- **`debt_to_equity`**: to analyze the ratio between the company's total debt and its equity, which helps to understand the level of indebtedness and financial risk.
- **`trailing_pe`**: (Price to Earnings) to compare the current stock price with the earnings per share over the last 12 months, a classic metric for evaluating the price-earnings ratio.
- **`forward_pe`**: for projecting the stock price relative to expected earnings, providing a future estimate of the price-earnings ratio.
- **`price_to_book`**: for comparison between the market value of the company and its book value, which can signal stocks traded below their net asset value.
- **`earnings_quarterly_growth`**: to assess the company's earnings growth compared to the same quarter of the previous year, showing growth or decline trends.

These columns would be at the core of Buy and Hold analysis, as they reflect the fundamental aspects long-term investors generally seek: sustainable growth, financial solidity, and dividend return. The 
visualization of data filtered by these columns offers clear and precise insights, making the investment analysis process more efficient and well-founded.

## **Initial Setup**

### Install Packages

In [2]:
%pip install pandas -q
%pip install plotly -q

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


### Import libs

In [3]:
import os
import itertools
import pandas as pd
from pathlib import Path
import plotly.express as px
import plotly.graph_objects as go
import plotly.subplots as sp

### Pandas Config

In [4]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Create a file path default

In [5]:
file_path_book = str(Path(os.getcwd()).parent.parent.parent / "data/book")

## Companies

### Load data

In [6]:
df_fundamentals_book = pd.read_csv(file_path_book + "/fundamentals_book.csv")
df_fundamentals_book.head(5)

ticker                                          long_name  \
0  ABCB4.SA                              Banco ABC Brasil S.A.   
1  AGRO3.SA  BrasilAgro - Companhia Brasileira de Proprieda...   
2  RAIL3.SA                                          Rumo S.A.   
3  ALPA3.SA                                    Alpargatas S.A.   
4  ALPA4.SA                                    Alpargatas S.A.   

               sector                industry    market_cap  enterprise_value  \
0  Financial Services        Banks - Regional  4.265434e+09      1.477339e+10   
1  Consumer Defensive           Farm Products  2.466480e+09      2.912933e+09   
2         Industrials               Railroads  4.228882e+10      5.524305e+10   
3   Consumer Cyclical  Footwear & Accessories  5.309793e+09      6.482982e+09   
4   Consumer Cyclical  Footwear & Accessories  5.350758e+09      6.395236e+09   

   total_revenue  profit_margins  operating_margins  dividend_rate   beta  \
0   1.941779e+09         0.41576            0.38826           1.56  0.679   
1   1.249437e+09         0.21493            0.25031           3.21  0.432   
2   1.031746e+10         0.07639            0.33544           0.07  0.227   
3   4.022153e+09        -0.05671           -0.06434           0.40  0.571   
4   4.022153e+09        -0.05671           -0.06434           0.43  0.571   

         ebitda  trailing_pe  forward_pe     volume  average_volume  \
0  0.000000e+00     4.069768    4.706601    92300.0        747165.0   
1  2.648920e+08     9.450382    6.332481   298100.0        666692.0   
2  4.522541e+09    54.309525   21.723810  5733400.0      14644522.0   
3 -1.980000e+05     0.000000    0.000000     1100.0          3953.0   
4 -1.980000e+05     0.000000   14.555555  1132100.0       5605825.0   

   fifty_two_week_low  fifty_two_week_high  price_to_sales_trailing_12_months  \
0               15.85                21.99                           2.196663   
1               22.29                32.71                           1.974073   
2               16.21                24.44                           4.098764   
3                7.27                17.80                           1.320137   
4                6.81                22.51                           1.330322   

   fifty_day_average  two_hundred_day_average  trailing_annual_dividend_rate  \
0            19.3382                 18.14667                          1.550   
1            27.0106                 25.58635                          3.240   
2            22.5852                 20.95235                          0.066   
3             8.7146                  9.63540                          0.000   
4             8.3228                  9.27290                          0.000   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.080687      24.518       0.785138  7.774306e+09   
1                        0.132029      22.237       1.113460  3.838370e+08   
2                        0.002993       8.334       2.736981  7.656040e+09   
3                        0.000000       7.867       1.008008  4.142880e+08   
4                        0.000000       7.867       0.999110  4.142880e+08   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                35.162  1.829846e+10                      0.001   
1                 3.885  8.720750e+08                      6.801   
2                 4.132  2.184320e+10                      3.935   
3                 0.614  1.550341e+09                      0.000   
4                 0.614  1.550341e+09                      0.000   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0           0.003        0.00000         0.00000           0.01530   
1           0.671        0.25252         0.21201           0.03839   
2           0.121        0.34493         0.43834           0.04252   
3          -0.127        0.43246        -0.00005          -0.00910   
4          -0.127        0.43246   

In [7]:
df_fundamentals_numeric_cols = df_fundamentals_book.select_dtypes(include=["int", "number", "float64"])
df_fundamentals_numeric_cols.head(5)

market_cap  enterprise_value  total_revenue  profit_margins  \
0  4.265434e+09      1.477339e+10   1.941779e+09         0.41576   
1  2.466480e+09      2.912933e+09   1.249437e+09         0.21493   
2  4.228882e+10      5.524305e+10   1.031746e+10         0.07639   
3  5.309793e+09      6.482982e+09   4.022153e+09        -0.05671   
4  5.350758e+09      6.395236e+09   4.022153e+09        -0.05671   

   operating_margins  dividend_rate   beta        ebitda  trailing_pe  \
0            0.38826           1.56  0.679  0.000000e+00     4.069768   
1            0.25031           3.21  0.432  2.648920e+08     9.450382   
2            0.33544           0.07  0.227  4.522541e+09    54.309525   
3           -0.06434           0.40  0.571 -1.980000e+05     0.000000   
4           -0.06434           0.43  0.571 -1.980000e+05     0.000000   

   forward_pe     volume  average_volume  fifty_two_week_low  \
0    4.706601    92300.0        747165.0               15.85   
1    6.332481   298100.0        666692.0               22.29   
2   21.723810  5733400.0      14644522.0               16.21   
3    0.000000     1100.0          3953.0                7.27   
4   14.555555  1132100.0       5605825.0                6.81   

   fifty_two_week_high  price_to_sales_trailing_12_months  fifty_day_average  \
0                21.99                           2.196663            19.3382   
1                32.71                           1.974073            27.0106   
2                24.44                           4.098764            22.5852   
3                17.80                           1.320137             8.7146   
4                22.51                           1.330322             8.3228   

   two_hundred_day_average  trailing_annual_dividend_rate  \
0                 18.14667                          1.550   
1                 25.58635                          3.240   
2                 20.95235                          0.066   
3                  9.63540                          0.000   
4                  9.27290                          0.000   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.080687      24.518       0.785138  7.774306e+09   
1                        0.132029      22.237       1.113460  3.838370e+08   
2                        0.002993       8.334       2.736981  7.656040e+09   
3                        0.000000       7.867       1.008008  4.142880e+08   
4                        0.000000       7.867       0.999110  4.142880e+08   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                35.162  1.829846e+10                      0.001   
1                 3.885  8.720750e+08                      6.801   
2                 4.132  2.184320e+10                      3.935   
3                 0.614  1.550341e+09                      0.000   
4                 0.614  1.550341e+09                      0.000   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0           0.003        0.00000         0.00000           0.01530   
1           0.671        0.25252         0.21201           0.03839   
2           0.121        0.34493         0.43834           0.04252   
3          -0.127        0.43246        -0.00005          -0.00910   
4          -0.127        0.43246        -0.00005          -0.00910   

   return_on_equity  gross_profits  total_assets_approx  asset_turnover  \
0           0.15680   1.973086e+09         7.774306e+09        0.249769   
1           0.12170   3.155040e+08         3.838370e+08        3.255124   
2           0.05163   3.146360e+09         7.656040e+09        1.347623   
3          -0.04153   1.968303e+09         4.142880e+08        9.708591   
4          -0.04153   1.968303e+09         4.142880e+08        9.708591   

   earnings_growth_rate  dividend_payout_ratio        equity  debt_to_equity  \
0                   0.1          155000.000000 -1.052416e+10       -1.738710   
1             

In [52]:
df_top_companies = df_fundamentals_book.nlargest(18, ['market_cap', 'dividend_rate', 'trailing_annual_dividend_yield', 'trailing_annual_dividend_rate', 'debt_to_equity', 'return_on_equity', 'trailing_pe', 'forward_pe', 'price_to_book', 'earnings_quarterly_growth']).drop_duplicates(subset='long_name', keep='first').reset_index(drop=True)
df_top_companies

ticker                                         long_name  \
0   PETR3.SA              Petróleo Brasileiro S.A. - Petrobras   
1   VALE3.SA                                         Vale S.A.   
2   ITUB4.SA                        Itaú Unibanco Holding S.A.   
3   ABEV3.SA                                        Ambev S.A.   
4   BPAC5.SA                            Banco BTG Pactual S.A.   
5   BBDC4.SA                               Banco Bradesco S.A.   
6   BBAS3.SA                              Banco do Brasil S.A.   
7   WEGE3.SA                                          WEG S.A.   
8   ITSA4.SA                                       Itaúsa S.A.   
9   ELET3.SA  Centrais Elétricas Brasileiras S.A. - Eletrobrás   
10  VIVT3.SA                            Telefônica Brasil S.A.   

                    sector                          industry    market_cap  \
0                   Energy              Oil & Gas Integrated  4.832119e+11   
1          Basic Materials  Other Industrial Metals & Mining  2.776947e+11   
2       Financial Services                  Banks - Regional  2.486832e+11   
3       Consumer Defensive               Beverages - Brewers  2.007233e+11   
4       Financial Services                   Capital Markets  1.483266e+11   
5       Financial Services                  Banks - Regional  1.423388e+11   
6       Financial Services                  Banks - Regional  1.404060e+11   
7              Industrials    Specialty Industrial Machinery  1.384426e+11   
8              Industrials                     Conglomerates  8.658098e+10   
9                Utilities             Utilities - Renewable  7.929263e+10   
10  Communication Services                  Telecom Services  7.645024e+10   

    enterprise_value  total_revenue  profit_margins  operating_margins  \
0       7.293640e+11   5.815630e+11         0.26889            0.39067   
1       3.399071e+11   2.064140e+11         0.27585            0.33672   
2       7.077875e+11   1.189380e+11         0.26661            0.35896   
3       1.888459e+11   8.271054e+10         0.17263            0.18666   
4       4.345411e+10   2.775135e+10         0.30292            0.34960   
5       5.421110e+11   7.289319e+10         0.20718            0.22324   
6       9.574406e+11   9.532893e+10         0.34830            0.45908   
7       1.303664e+11   3.175831e+10         0.15827            0.20575   
8       9.397528e+10   7.807000e+09         1.69976            0.20113   
9       1.201184e+11   3.551173e+10         0.04797            0.42603   
10      8.845781e+10   5.031185e+10         0.08991            0.16545   

    dividend_rate   beta        ebitda  trailing_pe  forward_pe      volume  \
0            9.28  1.097  2.910870e+11     2.914199    4.821608   2993300.0   
1            3.84  0.821  8.536400e+10     5.319043    5.421849   4387500.0   
2            1.50  0.499  0.000000e+00     8.557632    6.867500  10480300.0   
3            0.76  0.604  2.225154e+10    14.166667   12.750000   8618800.0   
4            0.35  1.291  0.000000e+00    11.260274    0.000000      2600.0   
5            0.23  0.675  0.000000e+00    10.649254    6.150862  14450200.0   
6            0.94  0.803  0.000000e+00     3.770115    3.805105   1529200.0   
7            0.32  0.513  6.480928e+09    26.829268   24.444445   8227600.0   
8            0.74  0.647  2.240000e+09     6.740458    5.553459   2919900.0   
9            0.22  0.629  1.274860e+10    34.700000    8.361445    867000.0   
10           2.58  0.061  1.510794e+10    17.029520   13.613569    518500.0   

    average_volume  fifty_two_week_low  fifty_two_week_high  \
0       13234123.0               23.61                41.86   
1       25978987.0               61.00                98.29   
2       24528203.0               22.62                31.29   
3       24680549.0               12.33                16.88   
4           6184.0                4.55                 9.08   
5       33278303.0               12.39                20.56 

To promote effective portfolio diversification, we refine the dataset to include only one ticker per company, choosing the 10 companies with the best financial metrics.

### Companies Analisys

#### Outlier Analysis

In [48]:
df_top_companies_cols_num = df_top_companies.select_dtypes(include= ['number', 'float64'])

top_companies_cols = df_top_companies_cols_num.columns
num_columns = len(top_companies_cols)
num_rows = (num_columns + 2) // 3

subplot_titles = [str(col) for col in top_companies_cols]

fig = sp.make_subplots(rows=num_rows, cols=3, subplot_titles=subplot_titles)

for i, column in enumerate(top_companies_cols, start=1):
    row = (i - 1) // 3 + 1
    col = (i - 1) % 3 + 1

    trace = go.Box(y=df_top_companies[column], name=column, 
                   marker_color='lightseagreen', boxpoints='outliers', 
                   jitter=0.7, hoverinfo='y+text', 
                   text=(df_top_companies['ticker']+ ' - ' + df_top_companies['long_name']))

    fig.add_trace(trace, row=row, col=col)

fig.update_layout(title_text='Boxplot of Numerical Variables', height=300*num_rows, showlegend=False, template='plotly_dark')
fig.show()

**Petróleo Brasileiro S.A. - Petrobras: (PETR3.SA)**

- **`market_cap`**:  Petrobras has an atypical market capitalization of **R$ 483,211,900,000**, reflecting its significant size in the energy sector.
- **`total_revenue`**: **R$ 581.563 billion**. This metric reflects the total amount of revenue generated by Petrobras in its financial operations.
- **`dividend_rate`**: The dividend rate of Petrobras at **9.28%** of the current stock value is relatively high, which may be seen as positive for investors seeking to receive significant dividends.
- **``**
- **`trailing_annual_dividend_rate`**: O valor incomum de **9.65%** para a Petrobras pode ser considerado um outlier, indicando uma taxa de dividendos significativamente alta em comparação com outras empresas.
- **`trailing_annual_dividend_yield`**: The value of **0.24%** for Petrobras does not appear to be an outlier, as it represents a dividend rate relative to the stock price within a reasonable range when compared to other companies in the same sector.
- **`gross_profits`**: The field is **R$ 334.1 billion**, representing the company's total revenue before the deduction of operating expenses, taxes, and other charges.
- **`roce`**: Return on Capital Employed is **0.38171**, indicating the company's efficiency in generating a return on invested capital. The higher the value, the better the efficiency in using its financial resources.

**Itaúsa S.A. (ITSA4.SA)**

- **`profit_margins`**:The field for Itaúsa represents the company's profit margin, which is **1.69976%**. This metric indicates the percentage of profit earned by Itaúsa in relation to its total revenue. That margin could be considered relatively low.
- **`price_to_sales_trailing_12_months`**: The value of 11.09017 for Itaúsa indicates that the market is willing to pay **11.09x**  the company's sales revenue over the past 12 months for each share, which may reflect optimism regarding the company's financial performance.

**Banco BTG Pactual S.A. (BPAC5.SA)**

- **`beta`**:The beta value of 1.291 for Banco BTG Pactual S.A. indicates a relatively high level of volatility in its shares in relation to the market.

In [41]:
fig = px.bar(top_companies, x='long_name', y=['profit_margins', 'operating_margins'],
             title='Margens de Lucro das Top 10 Empresas por Market Cap',
             labels={'value': 'Margem', 'variable': 'Tipo de Margem'},
             barmode='group')

fig.update_layout(xaxis={'categoryorder':'total descending'}, xaxis_title='Empresa', yaxis_title='Margem')
fig.show()
